### PR TITLE
Enforce reload before publishing

### PR DESCRIPTION
**Pull request by Agrair**
### What is it?
Does what it says. Solves #1050, with very little code :^)

### Other implementations
To be honest, I'm not sure. This implementation definitely works but I dunno if it fits your guys' idea of good practice. In the future, maybe a new translation key that reads "Reload Before Publishing" or just "Reload" (which would fit a lot better actually) that can be used instead just reusing the "Reload Mods" one might be nice, but I think any modder who reads it should probably be able to figure it out based on context anyway.

### CODE REVIEW

File: patches/tModLoader/Terraria/ModLoader/UI/UIModSourceItem.cs [hunk 1]
```diff
@@ -40,7 +40,15 @@ internal class UIModSourceItem : UIPanel
...
 
-			string addendum = Path.GetFileName(mod).Contains(" ") ? $"  [c/FF0000:{Language.GetTextValue("tModLoader.MSModSourcesCantHaveSpaces")}]" : "";
+			bool isLoaded = ModLoader.GetMod(mod) != null;
+
+			string addendum = "";
+			if (Path.GetFileName(mod).Contains(" ")) {
+				addendum = $"  [c/FF0000:{Language.GetTextValue("tModLoader.MSModSourcesCantHaveSpaces")}]";
+			}
+			else if (builtMod != null && !isLoaded) {
+				addendum = $"  [c/FF0000:{Language.GetTextValue("tModLoader.ModReloadRequired")}]";
```
Comment (Agrair) [inline patches/tModLoader/Terraria/ModLoader/UI/UIModSourceItem.cs:50]: If you're reviewing my PR and aren't sure what I meant by reusing the "Reload Required" translation, here it is. But I think it should be fine.

**Comment by Agrair** [comment]
Actually maybe I should have made an isolated "Reload" button, and just scrap the "Reload Required' header.

**Comment by Jofairden** [comment]
From my personal understanding, the idea was just that the mod would cause a reload once you press publish, and after a successful reload process the publishing process can begin. The reload process will throw an exception in case of failure either way, which should stop the publish process from triggering.

**Comment by Agrair** [comment]
Hmm ok. Looking it over it's bugged anyway because if the mod is already enabled, it'll crash when you press Reload. (Enabled should be set only if it's false because it calls .Add on Mods)

**Comment by Agrair** [comment]
So I did Jof's suggestion. It looks a bit like a workaround, but it works like a charm. And sorry for all the force-pushing, now that I think about it that was probably unnecessary, but I just *really* wanted the PR to display a single commit. Feel free to look at the other ones though.

**Comment by Chicken-Bones** [comment]
Okay, this is sweet. Now we just need to enable a flag here that turns memory leak warnings into errors.
Happy to accept as is if you'd like

**Comment by Mirsario** [comment]
> Now we just need to enable a flag here that turns memory leak warnings into errors.

You mean throwing errors on every remaining static field value? That's gonna get ugly way too quick. Better just assume that we'll finally move our asses to .NET Core eventually, for that assembly unloading. Or, just come up with a way of knowing what static constructors were called and which weren't. That way we'll be able to null out everything on unload, and reinitialize things on reload.

**Comment by Agrair** [comment]
Hmm. Obviously there’s a bunch to be desired with enforcing not only successful loads but also proper unload mechanics, but I agree it’ll probably be messy. If you’re asking me if I’m ok with it getting merged, then yeah I am. I’m gonna take a peek at the GC stuff and see if I can put what Mirsar said into practice, but in all likelihood that’s either not going to end up being committed at all or just long after this PR should have been merged in the first place.

**Pull request merged**